### PR TITLE
Allow user to disable the automatic formatting of checkpoint file names.

### LINKED ISSUE #4012: ModelCheckpoint filename unable to use metrics that contain a slash
## 🐛 Bug

ModelCheckpoint is unable to save filenames that reference a metric with a slash in their name. I use grouped metrics for tensorboard, and would like to save my files containing my loss: `val/loss`. However, ModelCheckpoint uses `os.path.split`, which splits the file name: https://github.com/PyTorchLightning/pytorch-lightning/blob/6ac0958166c66ed599c96737b587232b7a33d89e/pytorch_lightning/callbacks/model_checkpoint.py#L258

If I try to use 
```python
ModelCheckpoint("root/dir/{epoch}_{val/loss:.5f}")
```

The above evaluates to 
```python
self.dirpath = "root/dir/{epoch}_{val" 
self.filename = "loss:.5f}"
```

This inevitably causes failure when attempting to format the output path.

### To Reproduce

As above, log a metric with a slash, then use it in model checkpoint output

#### Code sample

```python
class Module(pl.LightningModule):
...
    def validation_step(self, batch, batch_idx):
        x, y = batch
        logits = self.forward(x)
        loss = self.loss_fn(logits, y)
        self.log('val/loss', loss, on_epoch=True)
        return loss

...
def main():
    trainer = pl.Trainer(checkpoint_callback=ModelCheckpoint("{epoch}_{val/loss:.5f}"))
```

### Expected behavior

Split only along file path boundaries, ignoring variable names yet-to-be-formatted.
Per the previous example, we'd expect:
```python
self.dirpath = "root/dir" 
self.filename = "{epoch}_{val/loss:.5f}"
```

### Environment

* CUDA:
	- GPU:
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0
	- pytorch-lightning: 0.10.0
	- tqdm:              4.50.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.5
	- version:           #1 SMP Fri Sep 4 14:19:36 UTC 2020

### PR DESCRIPTION
**Pull request by maxfrei750**
## What does this PR do?

* adds parameter/attribute `auto_insert_metric_name` to `callbacks.ModelCheckpoint`
* adds parameter `auto_insert_metric_name` to `callbacks.ModelCheckpoint._format_checkpoint_name`
* if `auto_insert_metric_name=False`, then metric names are not added to checkpoint filename automatically
* default is `auto_insert_metric_name=True`, so that the changes are backwards compatible
* fixes #4012

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
:+1:

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by maxfrei750** [comment]
> LGTM !

I agree.

@Borda Since you are already involved in this PR, would you mind doing a review?

@tchaton @Borda Please suggest and ping a third reviewer.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -131,6 +131,15 @@ class ModelCheckpoint(Callback):
         ...     filename='sample-mnist-{epoch:02d}-{val_loss:.2f}'
         ... )
 
+        # save epoch and val_loss in name, but specify the formatting yourself (e.g. to avoid problems with Tensorboard
+        # or Neptune, due to the presence of characters like = or /)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:135]:
```suggestion
        # or Neptune, due to the presence of characters like '=' or '/')
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -41,6 +44,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `trainer.evaluating` to return `True` if validating or testing ([#4945](https://github.com/PyTorchLightning/pytorch-lightning/pull/4945))
 
 
+- Renamed `pytorch_lightning.callbacks.swa` to `pytorch_lightning.callbacks.stochastic_weight_avg` ([#6259](https://github.com/PyTorchLightning/pytorch-lightning/pull/6259))
+
+
```
Comment (carmocca) [inline CHANGELOG.md:49]: This is duplicated

```suggestion
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -131,6 +131,15 @@ class ModelCheckpoint(Callback):
...
+        # saves a file like: my/path/sample-mnist-epoch02-val_loss0.32.ckpt
+        >>> checkpoint_callback = ModelCheckpoint(
+        ...     monitor='val/loss',
+        ...     dirpath='my/path/',
+        ...     filename='sample-mnist-epoch{epoch:02d}-val_loss{val/loss:.2f}'
+        ... )
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:141]: Isn't this missing `auto_insert_metric_name=Fale`?

Comment (maxfrei750) [reply pytorch_lightning/callbacks/model_checkpoint.py]: Good catch. Fixed in 278cef7.

**Review by carmocca — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -367,7 +380,10 @@ def _format_checkpoint_name(
             metrics.update({"epoch": epoch, 'step': step})
             for group in groups:
                 name = group[1:]
-                filename = filename.replace(group, name + "={" + name)
+
+                if auto_insert_metric_name:
+                    filename = filename.replace(group, name + "={" + name)
```
Comment (talregev) [inline pytorch_lightning/callbacks/model_checkpoint.py:385]: filename = filename.replace('/', '_')

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:385]: Hey @talregev 
Thanks for the comment. 
Could you open an issue about it, describing the need for this?
